File: library/java/eventdispatcher/EventDispatcher.java
package com.jaronho.sdk.library.eventdispatcher;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Author:  jaron.ho
 * Date:    2017-02-07
 * Brief:   EventDispatcher
 */

public class EventDispatcher {
    public static abstract class Handler {
        public abstract void onCallback(Object param);
    }

    private ConcurrentHashMap<Object, List<Handler>> mEventHandlerMap = new ConcurrentHashMap<>();

    public void subscribe(Object eventId, Handler handler) {
        List<Handler> handlers = mEventHandlerMap.get(eventId);
        if (null == handlers) {
            handlers = new ArrayList<>();
            mEventHandlerMap.put(eventId, handlers);
        }
        if (!handlers.contains(handler)) {
            handlers.add(handler);
        }
    }

    public void unsubscribe(Object eventId, Handler handler) {
        if (null == eventId) {
            mEventHandlerMap.clear();
        } else if (null == handler) {
            mEventHandlerMap.remove(eventId);
        } else {
            List<Handler> handlers = mEventHandlerMap.get(eventId);
            if (null != handlers) {
                handlers.remove(handler);
                if (handlers.isEmpty()) {
                    mEventHandlerMap.remove(eventId);
                }
            }
        }
    }

    public void post(Object eventId, Object param) {
        List<Handler> handlers = mEventHandlerMap.get(eventId);
        if (null != handlers) {
            for (Handler handler : handlers) {
                if (null != handler) {
                    handler.onCallback(param);
                }
            }
        }
    }
}

File: alone/c#/eventdispatcher/EventDispatcher.cs
/*--------------------------------------------------------------------
 * Author:	jaron.ho
 * Date:	2016-11-01
 * Brief:	event dispathcer system
--------------------------------------------------------------------*/
using System.Collections;
using System.Collections.Generic;
//--------------------------------------------------------------------
public class EventDispatcher {
	private Hashtable mEventHandlerMap = new Hashtable();

	public delegate void EventHandler(System.Object obj);

	public void subscribe(System.Object eventId, EventHandler handler) {
		List<EventHandler> handlers = (List<EventHandler>)mEventHandlerMap[eventId];
		if (null == handlers) {
			handlers = new List<EventHandler>();
			mEventHandlerMap.Add(eventId, handlers);
		}
		handlers.Add(handler);
	}

	public void unsubscribe(System.Object eventId, EventHandler handler = null) {
		List<EventHandler> handlers = (List<EventHandler>)mEventHandlerMap[eventId];
		if (null == handlers) {
			return;
		}
		if (null == handler) {
			mEventHandlerMap.Remove(eventId);
		} else {
			handlers.Remove(handler);
			if (0 == handlers.Count) {
				mEventHandlerMap.Remove(eventId);
			}
		}
	}

	public void post(System.Object eventId, System.Object param = null) {
		List<EventHandler> handlers = (List<EventHandler>)mEventHandlerMap[eventId];
		if (null == handlers) {
			return;
		}
		handlers.ForEach(delegate(EventHandler handler) {
			handler(param);
		});
	}

	public void post(int eventId) {
		post(eventId, null);
	}
}
//--------------------------------------------------------------------
public class EventCenter {
	private static EventDispatcher mEventDispatcher = new EventDispatcher();
	public static void subscribe(System.Object eventId, EventDispatcher.EventHandler handler) {
		mEventDispatcher.subscribe(eventId, handler);
	}
	public static void unsubscribe(System.Object eventId, EventDispatcher.EventHandler handler = null) {
		mEventDispatcher.unsubscribe(eventId, handler);
	}
	public static void unsubscribe(System.Object eventId) {
		mEventDispatcher.unsubscribe(eventId, null);
	}
	public static void post(System.Object eventId, System.Object param = null) {
		mEventDispatcher.post(eventId, param);
	}
	public static void post(System.Object eventId) {
		mEventDispatcher.post(eventId);
	}
}
//--------------------------------------------------------------------
